### ISSUE TITLE
sparse multilingual models scaling experiments

### ISSUE THREAD

**Issue by swag2198** [OpenEuroLLM/Taskboard#216]
This is a parent issue to keep track of everything related to sparse (MoE) multilingual models scaling law experiments.
- Model architecture prepared in this [sheet](https://docs.google.com/spreadsheets/d/17LXPs6O9_6cL7P2KVEJF8LTy5zhy0qLNXEVGRCYOzhY/edit?gid=500956383#gid=500956383)
- Abhash's dashboard for english runs: https://abhash-er.github.io/moe-analysis-dashboard/
- Updated configs on the `exp_moe_scaling` branch: [models](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_moe_scaling/config/backend/megatron/multilingual_scaling/sparse_models), [sweeps](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_moe_scaling/config/sweep/multilingual_scaling/sparse_scaling), main [configs](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_moe_scaling/config/experiments/multilingual_scaling/sparse_scaling) to run with run_autoexp.

Related issues:
- #202 (caution: the models trained here show significant degradation in MoE layers, routing collapse etc., see [comment](https://github.com/OpenEuroLLM/Taskboard/issues/216#issuecomment-4715693324) below)
- #165 (small sweep limited to only 10B tokens)

**Comment by swag2198** [comment]
## TODOs before starting the runs:
- [x] Prepare configs @aaronkl @swag2198: we are going with increasing sparsity configs, same active parameters as the dense multilingual models but with fine-grained experts
- [x] Initial throughput testing and parallelism settings with configs on Leonardo @swag2198 
- [x] Initial throughput testing and parallelism settings with configs on JUPITER @dharmsen 
- [x] Analysis of existing checkpoints to finalize load balancing settings @abhash-er @luukkonenr 
- [x] Figure out storage situation on Leonardo
- [ ] Ensure everyone has updated progress tracker to track status of runs


## Center training runs 
- [ ] 0.1B @dharmsen 
- [ ] 0.2B @abhash-er 
- [ ] 0.4B @swag2198 
- [ ] 0.9B @aaronkl

**Comment by swag2198** [comment]
## Initial throughput testing results on Leonardo:

also included the dense model TFLOPs for comparison.

Model | Sparsity (1 - K/E) | Parallelism settings | TFLOPs/s/GPU | Tokens/s/GPU | Max dense TFLOPs/s/GPU
-- | -- | -- | -- | -- | --
0.1B_ne / 0.24B | 4/8 ⇒ 50% | EP=2 | 48-58 | 28k-35k | 60-70
0.2B_ne / 0.74B | 4/16 ⇒ 75% | EP=2 | 36-53 | 12k-18k | 60-75
0.4B_ne / 2.56B | 4/32 ⇒ 87.5% | MBS=2, EP=4 | 40 | 7k | 60-75
0.9B_ne / 10.18B | 4/64 ⇒ 93.7% | **MBS=2, EP=4, PP=2** | 30-32 | 2.9k-3.5k | 65-67
0.9B_ne / 10.18B | 4/64 ⇒ 93.7% | MBS=1 or 2, EP=4 | 22-26 | 2.5k-3k |  


## Throughput testing results on JUPITER #229 

from @dharmsen 

| model | nodes | mbs | gbs | tflops/s/gpu | tokens/s/gpu |
| --- | --- | --- | --- | --- | --- |
| 0.1 | 1 | 1 | 64 | 24 | 24644 |
| 0.1 | 1 | 2 | 64 | 43 | 44488 |
| 0.1 | 1 | 4 | 64 | 76 | 77509 |
| 0.1 | 1 | 8 | 64 | 114 | 116416 |
| 0.2 | 1 | 1 | 64 | 16 | 8720 |
| 0.2 | 1 | 2 | 64 | 57 | 29996 |
| 0.2 | 1 | 4 | 64 | 106 | 56143 |
| 0.2 | 1 | 8 | 64 | 146 | 76961 |
| 0.4 | 1 | 1 | 64 | 40 | 9295 |
| 0.4 | 1 | 2 | 64 | 80 | 18367 |
| 0.4 | 1 | 4 | 64 | 145 | 33391 |
| 0.4 | 1 | 8 | 64 | 162 | 37246 |

**Comment by swag2198** [comment]
## English runs problems

The 130M active "english data" MoE models (#202) were trained with the following _recipe_:
- sweep over lr x bsz grid: [5.e-4, 1.e-3, 2.e-3, 4.e-3, 8.e-3] x [1024, 512, 256]
- **optimizer** [config](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_swagatam/config/experiments/swagatam/moe_130M_active_configs/training/test_moe_130M_300BT_bsz_1024.yaml#L48): warmup_steps = 2000, (beta1, beta2) = (0.9, 0.95), min_lr = 1e-5
- **moe** settings: topk=2 always, grid over total experts 8 to 64, aux_loss = 1e-3 minibatch level, z_loss = none.

The above recipe unfortunately produced collapsed models as suggested by multiple metrics in the [dashboard](https://abhash-er.github.io/moe-analysis-dashboard/).
- we also tried using global_batch_aux_loss (with aux loss coeff 1e-3) and deepseek_bias, it did not seem to help! But it may be because the coeff was too low?
- what did seem to **help** was to bump up the aux loss coeff to 1e-2 (which is prevalent in literature).

Next steps to try as suggested by @luukkonenr @abhash-er @aaronkl :
1) identify one of the recipes to replicate one of the crashing router runs. How long will it take? How many GPUh is required? How much can we afford? 
2) Execute a sequence / grid of experiments with varying params such as aux loss, z-loss, learning rate etc. to find at least some stable configuration(s). 
3) Move from there. 
4) Also look at open source checkpoints, candidates: OLMoE (1B-7B), FlameMoE and StepLaw?

### Interventions to try:
Looking at the dashboard, it seems instabilities kick in quite early ~20B tokens, and from then on just becomes worse (few experts share most of the load), so we can save a lot on compute by training models upto 50B or 60B tokens.
- higher warm up
- aux_loss coeff (1e-3, 5e-3, 1e-2) x aux_loss strategies grid (micro batch, global batch)
-

**Comment by abhash-er** [comment]
For the reference of the issue- I'd briefly lay the steps for getting the metrics from a checkpoint folder. Overall the structure remain simple as mentioned in the earlier [issue](https://github.com/OpenEuroLLM/Taskboard/issues/164). But I'd recommend to add it as a config in `autoexp` so that you avoid writing so many CLI parameters. See [here](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_abhash/config/experiments/abhash/moe_metrics_120BT_global_batch_aux.yaml) for an example of how you can add a run as a config.

In the yaml config you need to specify the metadata parameters. You may copy the whole file and change some relevant metadata pertaining to the checkpoint you want to generate metrics for. 

```yaml
metadata:
  bsz: 256
  nexp: 8
  lr: "0.001"
  algo: global_batch_aux

  results_root: /leonardo_work/OELLM_prod2026/users/shaldar0/oellm-autoexp/results
  analysis_root: results/moe_analysis_${metadata.algo}
  idx_data: /leonardo_work/OELLM_prod2026/datasets/Nemotron-cc-2024-HQ-LUMI-sample-valid/high-all
  vocab_file: /leonardo_work/OELLM_prod2026/models/EleutherAI/gpt-neox-20b/vocab.json
  merge_file: /leonardo_work/OELLM_prod2026/models/EleutherAI/gpt-neox-20b/merges.txt

  grid_dir: ${metadata.results_root}/moe_no_GQA_${metadata.algo}_bsz${metadata.bsz}_lr_sparsity_grid_A130M_120BT
  run_tag: moe_no_GQA_${metadata.algo}_nexp_${metadata.nexp}_lr${metadata.lr}_gbsz${metadata.bsz}_seed1234
  stable_path: ${metadata.grid_dir}/${metadata.run_tag}_stable/checkpoints
  decay_path: ${metadata.grid_dir}/${metadata.run_tag}_decay120BT/checkpoints
  out_dir: ${metadata.analysis_root}/bsz${metadata.bsz}/nexp_${metadata.nexp}/lr${metadata.lr}/120BT
```

You can change the `results_root`, `grid_dir`, `stable_path`, `decay_path` for the checkpoint you want. Other metadata can be changed to your liking as well! You can run for the whole grid with one config by sweeping over all the metadata for batch size, num_experts etc. See [here](https://github.com/OpenEuroLLM/oellm-autoexp/blob/exp_abhash/config/experiments/abhash/moe_metrics_120BT.yaml) for an example.

Set relevant env variables- 
```
export SLURM_ACCOUNT="${SLURM_ACCOUNT:-OELLM_prod2026}"
export SLURM_PARTITION="${SLURM_PARTITION:-boost_usr_prod}"
export OELLM_CONTAINERS_DIR="${OELLM_CONTAINERS_DIR:-/leonardo_work/OELLM_prod2026/container_images}"
export HF_HOME="${HF_HOME:-/tmp}"
```  

then run, 
```python
python scripts/run_autoexp.py --config-name=YOUR_CONFIG
```

_Note: It is in development and the way things are run may change in future!_

**Comment by JeniaJitsev** [comment]
z-loss should be set to a non-zero value - it is used in most contemporary works, in our case it is deactivated. Might be one of the reasons router is collapsing.

**Comment by anni-moisala** [comment]
Throughput on Lumi is a bit less dreadful, same 1.7 active param config as in https://github.com/OpenEuroLLM/Taskboard/issues/128

### Throughput (TFLOP/s/GCD) 1-128 nodes

| Experiment                               |       1 |       2 |       4 |       8 |      16 |      32 |      64 |     128 |
| ---------------------------------------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- |
| qwen3-1.7B_tuner_on                      |  71.12~ |  70.34~ |  69.26~ |  68.31  |  66.95  |  63.90  |  57.78  |  41.73  |
| qwen3-10B_A1.7B_tuner_on                 |  48.40~ |  48.06~ |  47.63~ |  46.68~ |  45.04  |  38.51~ |  34.99  |  26.12  |

tuner refers to https://github.com/sfantao/rccl-tuner, here's a comparison with and without (it kicks in at 16 nodes)

<img width="1184" height="731" alt="Image" src="https://github.com/user-attachments/assets/7a353970-7690-475e-a46a-dc32e733dc8c" />

Will run the current scaling law configs next

**Comment by anni-moisala** [comment]
Throughput testing results on LUMI for autoexp [sparse model configs](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_moe_scaling/config/backend/megatron/multilingual_scaling/sparse_models) and [dense](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_moe_scaling/config/backend/megatron/multilingual_scaling/models) for comparison

All with GBS=64 and NODES=1

| Model | Sparsity | MBS | Parallelism settings | TFLOPS/s/GCD | Tokens/s/GCD | Dense TFLOPS/s/GCD | (Sparse TFLOPS/s/GCD) / (Dense TFLOPS/s/GCD) × 100 |
|-------|----------|-----|-----------------------|--------------|--------------|--------------------|---------------------|
| 0.1B  | 50%      | 1   | EP=2                  | 25.7         | 15685        | 38.2               | 67.3                |
| 0.1B  | 50%      | 2   | EP=2                  | 33.2         | 20287        | 42.8               | 77.6                |
| 0.1B  | 50%      | 4   | EP=2                  | 36.6         | 22379        | 44.2               | 82.8                |
| 0.2B  | 75%      | 1   | EP=2                  | 31.4         | 10905        | 48.4               | 64.9                |
| 0.2B  | 75%      | 2   | EP=2                  | 38.5         | 13373        | 51.8               | 74.3                |
| 0.2B  | 75%      | 4   | EP=2                  | 42.4         | 14730        | 53.8               | 78.8                |
| 0.4B  | 87.5%    | 1   | EP=4                  | 34.0         | 6004         | 56.5               | 60.2                |
| 0.4B  | 87.5%    | 2   | EP=4                  | 37.6         | 6632         | 59.6               | 63.1                |
| 0.4B  | 87.5%    | 4   | EP=4                  | oom          | oom          | oom                 |                     |
| 0.9B  | 93.75%   | 1   | EP=4                  | 29.4         | 3277         | 63.9               | 46.0                |
| 0.9B  | 93.75%   | 2   | EP=4                  | 30.8         | 3426         | 67.5               | 45.6                |
| 0.9B  | 93.75%   | 4   | EP=4                  | oom          | oom          | oom                |                     |

Notes: had to change `moe_router_dtype` from  "fp32" to "null" because of a hipBLASLt GEMM failure `Unable to find any suitable algorithms`
